codegen/src/next/mod.rs

```diff
@@ -0,0 +1,527 @@
+
+        route.to_tokens(tokens);
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct App {
+    endpoints: Vec<Endpoint>,
+}
+
+impl App {
+    pub(crate) fn from_file(file: &mut File) -> Self {
+        let endpoints = file
+            .items
+            .iter_mut()
+            .filter_map(|item| {
+                if let Item::Fn(item_fn) = item {
+                    Some(item_fn)
+                } else {
+                    None
+                }
+            })
+            .filter_map(Endpoint::from_item_fn)
+            .collect();
+
+        Self { endpoints }
+    }
+}
+
+impl ToTokens for App {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        let Self { endpoints } = self;
+
+        let app = quote!(
+            async fn __app(request: shuttle_next::Request<shuttle_next::body::BoxBody>,) -> shuttle_next::response::Response
+            {
+                use shuttle_next::Service;
+
+                let mut router = shuttle_next::Router::new()
+                    #(#endpoints)*;
+
+                let response = router.call(request).await.unwrap();
+
+                response
+            }
+        );
+
+        app.to_tokens(tokens);
+    }
+}
+
+pub(crate) fn wasi_bindings(app: App) -> proc_macro2::TokenStream {
+    quote!(
+        #app
+
+        #[cfg(not(test))]
+        #[no_mangle]
+        #[allow(non_snake_case)]
+        pub extern "C" fn __SHUTTLE_Axum_call(
+            logs_fd: std::os::wasi::prelude::RawFd,
+            parts_fd: std::os::wasi::prelude::RawFd,
+            body_read_fd: std::os::wasi::prelude::RawFd,
+            body_write_fd: std::os::wasi::prelude::RawFd,
+        ) {
+            use shuttle_next::body::{Body, HttpBody};
+            use shuttle_next::tracing_prelude::*;
+            use shuttle_next::Logger;
+            use std::io::{Read, Write};
+            use std::os::wasi::io::FromRawFd;
+
+            println!("inner handler awoken; interacting with fd={},{},{},{}", logs_fd, parts_fd, body_read_fd, body_write_fd);
+
+            // file descriptor 2 for writing logs to
+            let logs_fd = unsafe { std::fs::File::from_raw_fd(logs_fd) };
+
+            shuttle_next::tracing_registry()
+                .with(Logger::new(logs_fd))
+                .init(); // this sets the subscriber as the global default and also adds a compatibility layer for capturing `log::Record`s
+
+            // file descriptor 3 for reading and writing http parts
+            let mut parts_fd = unsafe { std::fs::File::from_raw_fd(parts_fd) };
+
+            let reader = std::io::BufReader::new(&mut parts_fd);
+
+            // deserialize request parts from rust messagepack
+            let wrapper: shuttle_next::RequestWrapper = shuttle_next::from_read(reader).unwrap();
+
+            // file descriptor 4 for reading http body into wasm
+            let mut body_read_stream = unsafe { std::fs::File::from_raw_fd(body_read_fd) };
```

Comment:
It was because streaming the bytes from the inner read_stream to the axum router consumed the stream, and so couldn't be used to write after. That didn't actually end up working, so it should be possible to use only one stream for the body now. On the runtime side I did also change it to drop the body write stream after it completes to signal EOF, but if we're reusing it we'd need to change it back to writing a `0`. I can look into this.

Reply:
There is also some method call that allows you to split the stream into a reader and writer. You can then safely drop the writer without needing to write the `0`

Reply:
Thanks for the tip! I found a way in #586